### ISSUE TITLE
View and edit modes

### ISSUE THREAD

**Issue by koke** [wordpress-mobile/gutenberg-mobile#204]
The design for 1.0 has two different editor modes: view and edit.

![view-edit-modes](https://user-images.githubusercontent.com/8739/47645212-9a849380-db70-11e8-9e08-0d1162025089.png)

@iamthomasbishop I have a few clarifying questions about these:

- How do you go from one to the other? From view you need to tap the edit button, or tapping on any block switches to edit mored? From edit, dismissing the keyboard goes back to view or is there another way?
- Does the "Save changes" button just go back to view mode or starts the publishing flow?

**Comment by iamthomasbishop** [comment]
@koke Happy to see this part being tackled :)

> How do you go from one to the other?

I hope this blueprint flow attached below helps clarify. Essentially:

* From `edit` mode, the `checkmark` icon will save the changes
* From `view` mode, you can double-tap on any part of the canvas, and the cursor should be placed as close to that point as possible. You can also tap on the Edit button on the bottom toolbar

![image](https://user-images.githubusercontent.com/1200432/47692913-8028d500-dbc5-11e8-9764-b97e4884f3d1.png)

> Does the "Save changes" button just go back to view mode or starts the publishing flow?

This saves changes (remote if possible, locally if necessary). The publishing flow should start when the user taps `Next`. I've also considered that we might want to include publishing directly from the `•••` menu in `Edit` mode, but I'd like to keep the publishing entry point to a single spot (via "Next" button on `View` mode).

**Comment by iamthomasbishop** [comment]
It's also worth noting that the flow above represents what it could look like on Android. This flow should be roughly the same on iOS, and look like this:

![image](https://user-images.githubusercontent.com/1200432/47693272-0a256d80-dbc7-11e8-9d78-1fe24b376389.png)

Notable difference is text labels instead of icons in most places, general flow is basically identical.

**Comment by koke** [comment]
That's clarifying, thanks. I only have some doubts about the double tap to edit, why not a single tap? What would happen instead when you single tap on a block?

**Comment by hypest** [comment]
To add to what Koke asked above, this seems to also be related to the focusing logic? Do focusing/unfocusing and view/edit relate to each other in some way?

**Comment by koke** [comment]
Considering we're reusing the apps' existing top bars for 1.0 and we're also not doing the new publish confirmation flow yet, this can wait until after the first beta.

**Comment by iamthomasbishop** [comment]
> That's clarifying, thanks. I only have some doubts about the double tap to edit, why not a single tap? What would happen instead when you single tap on a block?

@koke this behavior mimics other apps that utilize similar edit/view flows. IMO it adds just the right amount of friction between the modes, and feels natural. I'm not opposed to single-tap focusing, this is just utilizing a common pattern.

> Do focusing/unfocusing and view/edit relate to each other in some way?

@hypest I'm not exactly sure what you mean here, but the intention here is to separate writing vs. viewing contexts so that actions are clearer and more appropriate to the task. In terms of focusing/unfocusing – that's mainly left to edit mode, but you should be able to "enter" into focusing a single block as described above. Nothing groundbreaking, just following established patterns.

**Comment by hypest** [comment]
The web implementation of Gutenberg now has a "Unified Toolbar" mode which is ofc very similar to the mobile bottom toolbar, and where almost all block controls are living. Only the up/down buttons to move the block are outside that toolbar. I wonder if that toolbar mode can help clarify or remove the distinction between a view mode and an edit mode. I.e. can we try moving all controls to the bottom toolbar, remove the inline toolbar and find a place for the up/down functionality?

**Comment by shiki** [comment]
This looks really nice! 

I'm concerned about the _Save Changes_ (checkmark) button in _Edit_ mode. I would expect that to be a _Cancel_ button instead. Based on the funnel created by @maxme (see paCBwp-84-p2), about 84% of iOS users open the Editor only to eventually exit and discard their changes. But perhaps showing the _View_ mode first would fix this. 🤷‍♂

**Comment by koke** [comment]
> about 84% of iOS users open the Editor only to eventually exit and discard their changes

That funnel was using an unreliable event, the number is more like 33%.

> I would expect that to be a Cancel button instead

Keep in mind that Android already saves without asking on closing, and iOS is [headed in that direction](https://github.com/wordpress-mobile/WordPress-iOS/issues/11753).

I think when this was first explored, we didn't have any sort of autosaving (except on closing on Android), and I still think that's the right direction to go.

I don't know what the exact roadmap for autosaving is, but ideally, the editor would autosave periodically as you edit, and the checkmark/done button should just switch modes, encouraging the editor to save the content remotely if it hasn't already.

**Comment by iamthomasbishop** [comment]
> Ideally, the editor would autosave periodically as you edit, and the checkmark/done button should just switch modes, encouraging the editor to save the content remotely if it hasn't already.

Completely, 1000% agree – this was the intention. In cases where auto-save is not successful while editing, tapping the checkmark should at very least save the changes locally. My hope is that we can do all of this behind the scenes so that the user doesn't have to stress about it. 

Related to autosaving: We have heard in usability tests that users appreciate being able to see a small indication of some sort, so they have confidence their changes are saved – this is because people find writing on mobile, in general, to be unreliable.

**Comment by shiki** [comment]
Thank you for the answers, @koke and @iamthomasbishop. I agree with what y'all said. 👍

**Comment by SergioEstevao** [comment]
@iamthomasbishop do we still need to implement the edit/view modes? or are the design aboves outdated?

**Comment by iamthomasbishop** [comment]
@SergioEstevao Yes, we still need to implement this. We decided early on in the project to postpone this feature because it was expected to be a large one and could wait until we had the foundation built. Maybe now’s the time 🙂 

We haven’t substantially changed the designs/concept since the original proposal, so when we’re ready to prioritize it, I can do an updated blueprint and make some more detailed notes.
